swift/ql/test/library-tests/dataflow/dataflow/test.swift

```diff
@@ -631,3 +631,30 @@ func testSwap() {
     sink(arg: x) // $ SPURIOUS: flow=628
     sink(arg: y) // $ flow=628
 }
+
+func testArray() {
+    var arr1 = [1,2,3]
+    sink(arg: arr1[0])
+    arr1[1] = source()
+    sink(arg: arr1[0]) // $ flow=638
+    sink(arg: arr1)
+
+    var arr2 = [source()]
+    sink(arg: arr2[0]) // $ flow=642
+
+    var matrix = [[source()]]
+    sink(arg: matrix[0])
+    sink(arg: matrix[0][0]) // $ flow=645
```

Comment:
Does assignment to a multi-dimensional array work as well?
```
var matrix2 = [[1]]
matrix2[0][0] = source()
sink(arg: matrix2[0][0])
```

Reply:
No, it doesn't, and now that I look at the cases I based this off of I'm not sure we handle nested fields or tuples properly either...

Reply:
Mmm.  Then we should probably create an issue for this (in all content types) rather than trying to address it for just arrays here?

Reply:
After adding some tests it seems to work for nested fields... I'm not sure why, though.

Reply:
This is because of "reverse reads": if readStep(n2.(PostUpdateNode).getPreUpdate(); f, n1.(PostUpdateNode).getPreUpdate()) then the dataflow library generates storeStep(n1, f, n2). So if this doesn't work for array content then it's either because the post-update/pre-update isn't working, or because the readStep isn't working.

Reply:
I think it's actually that there's a step through an `InOutExpr` between the store and read step:
```
#  650|     getElement(11): [AssignExpr]  ... = ...
#  650|       getDest(): [SubscriptExpr] ...[...]
#  650|         getBase(): [InOutExpr] &...
#  650|           getSubExpr(): [SubscriptExpr] ...[...]
#  650|             getBase(): [InOutExpr] &...
#  650|               getSubExpr(): [DeclRefExpr] matrix2
```